In [28]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

## Stage 1

## What You Must Edit in the below Python Script

Before running the script, you only need to edit **3 things**:

1. Your Kaggle username
2. Your Kaggle API key
3. The folder path where you want the CSV to be saved

---

### Step 1: Paste Your Kaggle Username

Find this line in the Python script:

```python
os.environ["KAGGLE_USERNAME"] = "PASTE_YOUR_KAGGLE_USERNAME_HERE"
```

Replace it with your Kaggle username from `kaggle.json`.

Example:

```python
os.environ["KAGGLE_USERNAME"] = "soumo123"
```

---

### Step 2: Paste Your Kaggle API Key

Find this line in the Python script:

```python
os.environ["KAGGLE_KEY"] = "PASTE_YOUR_KAGGLE_KEY_HERE"
```

Replace it with your Kaggle API key from `kaggle.json`.

Example:

```python
os.environ["KAGGLE_KEY"] = "abcd1234yourapikey"
```

Do not add extra spaces before or after the key.

---

### Step 3: Paste Your Intended Save Folder

Find this line in the Python script:

```python
save_folder = Path(r"PASTE_YOUR_INTENDED_FOLDER_PATH_HERE")
```

Replace it with the folder where you want the dataset to be downloaded and saved.

Example:

```python
save_folder = Path(r"C:\Users\YourName\ITSM-Agent-Capstone\data\raw")
```

The `r` before the path is important on Windows because it helps Python read backslashes correctly.

---

### Final Edited Example

After editing, this part of your script should look like this:

```python
os.environ["KAGGLE_USERNAME"] = "your_kaggle_username"
os.environ["KAGGLE_KEY"] = "your_kaggle_api_key"

save_folder = Path(r"C:\Users\YourName\ITSM-Agent-Capstone\data\raw")
```

---

### Important Security Warning

Do not upload your real Kaggle username and API key to any platform.

Before sharing your code publicly, change the values back to:

```python
os.environ["KAGGLE_USERNAME"] = "PASTE_YOUR_KAGGLE_USERNAME_HERE"
os.environ["KAGGLE_KEY"] = "PASTE_YOUR_KAGGLE_KEY_HERE"
```

Your Kaggle API key is private and should not be exposed.


In [29]:
"""
import os
import sys
import glob
import subprocess
from pathlib import Path

# 1. Auto-install required packages
subprocess.check_call([sys.executable, "-m", "pip", "install", "-q", "kaggle", "pandas"])

import pandas as pd
from kaggle.api.kaggle_api_extended import KaggleApi


# 2. Paste Kaggle API details from kaggle.json
os.environ["KAGGLE_USERNAME"] = "PASTE_YOUR_KAGGLE_USERNAME_HERE"
os.environ["KAGGLE_KEY"] = "PASTE_YOUR_KAGGLE_KEY_HERE"


# 3. User chooses intended save folder
save_folder = Path(r"PASTE_YOUR_INTENDED_FOLDER_PATH_HERE")
save_folder.mkdir(parents=True, exist_ok=True)


# 4. Download Kaggle dataset
dataset = "mithilkotawadekar/incident-event-log-dataset"

api = KaggleApi()
api.authenticate()

api.dataset_download_files(
    dataset=dataset,
    path=str(save_folder),
    unzip=True,
    force=True
)

print("Dataset downloaded and unzipped.")


# 5. Auto-detect CSV file
csv_files = glob.glob(str(save_folder / "**" / "*.csv"), recursive=True)

print("CSV files found:", csv_files)

if not csv_files:
    raise FileNotFoundError("No CSV file found after download.")

csv_path = csv_files[0]


# 6. Load CSV
df = pd.read_csv(csv_path, low_memory=False)

print(f"SUCCESS! Loaded {df.shape[0]} rows and {df.shape[1]} columns.")
print("Loaded CSV path:", csv_path)


# 7. Preview
df.head(6)

"""

'\nimport os\nimport sys\nimport glob\nimport subprocess\nfrom pathlib import Path\n\n# 1. Auto-install required packages\nsubprocess.check_call([sys.executable, "-m", "pip", "install", "-q", "kaggle", "pandas"])\n\nimport pandas as pd\nfrom kaggle.api.kaggle_api_extended import KaggleApi\n\n\n# 2. Paste Kaggle API details from kaggle.json\nos.environ["KAGGLE_USERNAME"] = "PASTE_YOUR_KAGGLE_USERNAME_HERE"\nos.environ["KAGGLE_KEY"] = "PASTE_YOUR_KAGGLE_KEY_HERE"\n\n\n# 3. User chooses intended save folder\nsave_folder = Path(r"PASTE_YOUR_INTENDED_FOLDER_PATH_HERE")\nsave_folder.mkdir(parents=True, exist_ok=True)\n\n\n# 4. Download Kaggle dataset\ndataset = "mithilkotawadekar/incident-event-log-dataset"\n\napi = KaggleApi()\napi.authenticate()\n\napi.dataset_download_files(\n    dataset=dataset,\n    path=str(save_folder),\n    unzip=True,\n    force=True\n)\n\nprint("Dataset downloaded and unzipped.")\n\n\n# 5. Auto-detect CSV file\ncsv_files = glob.glob(str(save_folder / "**" / "*.csv"

In [30]:
import pandas as pd
csv_path = r"C:\Users\soumo\agy2-projects\ITSM-Agent-Capstone\data\raw\incident_event_log_dataset.csv" # It is my local "csv_path" , user must put their !!
df = pd.read_csv(csv_path)

(df.head(10))
print("-"*50)##########for space!!

print(f"Total number of Rows , Columns = {df.shape} \n")

print(f"ALL the names of columns in this dataset ----------> \n {df.columns.tolist()} \n")

print(f"Full structure of the dataset ------------> \n ")
df.info()



,number,incident_state,active,reassignment_count,reopen_count,sys_mod_count,made_sla,caller_id,opened_by,opened_at,...,notify,problem_id,change,vendor,caused_by,closed_code,resolved_by,resolved_at,closed_at,isParent
0,INC0000045,New,True,0,0,0,True,Caller 2403,Opened by 8,29-02-2016 01:16,...,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29-02-2016 11:29,05-03-2016 12:00,No
1,INC0000045,Resolved,True,0,0,2,True,Caller 2403,Opened by 8,29-02-2016 01:16,...,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29-02-2016 11:29,05-03-2016 12:00,No
2,INC0000045,Resolved,True,0,0,3,True,Caller 2403,Opened by 8,29-02-2016 01:16,...,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29-02-2016 11:29,05-03-2016 12:00,No
3,INC0000045,Closed,False,0,0,4,True,Caller 2403,Opened by 8,29-02-2016 01:16,...,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29-02-2016 11:29,05-03-2016 12:00,No
4,INC0000047,New,True,0,0,0,True,Caller 2403,Opened by 397,29-02-2016 04:40,...,Do Not Notify,?,?,?,?,code 5,Resolved by 81,01-03-2016 09:52,06-03-2016 10:00,No
5,INC0000047,Active,True,1,0,1,True,Caller 2403,Opened by 397,29-02-2016 04:40,...,Do Not Notify,?,?,?,?,code 5,Resolved by 81,01-03-2016 09:52,06-03-2016 10:00,No
6,INC0000047,Active,True,1,0,2,True,Caller 2403,Opened by 397,29-02-2016 04:40,...,Do Not Notify,?,?,?,?,code 5,Resolved by 81,01-03-2016 09:52,06-03-2016 10:00,No
7,INC0000047,Active,True,1,0,3,True,Caller 2403,Opened by 397,29-02-2016 04:40,...,Do Not Notify,?,?,?,?,code 5,Resolved by 81,01-03-2016 09:52,06-03-2016 10:00,No
8,INC0000047,Active,True,1,0,4,True,Caller 2403,Opened by 397,29-02-2016 04:40,...,Do Not Notify,?,?,?,?,code 5,Resolved by 81,01-03-2016 09:52,06-03-2016 10:00,No
9,INC0000047,Active,True,1,0,5,True,Caller 2403,Opened by 397,29-02-2016 04:40,...,Do Not Notify,?,?,?,?,code 5,Resolved by 81,01-03-2016 09:52,06-03-2016 10:00,No


--------------------------------------------------
Total number of Rows , Columns = (141712, 37) 

ALL the names of columns in this dataset ----------> 
 ['number', 'incident_state', 'active', 'reassignment_count', 'reopen_count', 'sys_mod_count', 'made_sla', 'caller_id', 'opened_by', 'opened_at', 'sys_created_by', 'sys_created_at', 'sys_updated_by', 'sys_updated_at', 'contact_type', 'location', 'category', 'subcategory', 'symptom', 'cmdb_ci', 'impact', 'urgency', 'priority', 'assignment_group', 'assigned_to', 'knowledge', 'u_priority_confirmation', 'notify', 'problem_id', 'change', 'vendor', 'caused_by', 'closed_code', 'resolved_by', 'resolved_at', 'closed_at', 'isParent'] 

Full structure of the dataset ------------> 
 
<class 'pandas.DataFrame'>
RangeIndex: 141712 entries, 0 to 141711
Data columns (total 37 columns):
 #   Column                   Non-Null Count   Dtype
---  ------                   --------------   -----
 0   number                   141712 non-null  str  
 1   inci

## Stage 2 Dataset Grain

In [31]:
print(f"Toatal number of event-level history rows = {len(df)}")

print(f"unique_incidents in this datset = {df['number'].nunique()}  \n")


rows_per_incident = df["number"].value_counts()
rows_per_incident.reset_index(name="Total_Rows/Ticket").head(10)
rows_per_incident.describe().reset_index(name="rows_per_incident")


# 1. Define the sample_incident & list of columns as lifecycle of one ticket:--------->
sample_incident = rows_per_incident.index[0]
lifecycle_cols = [
    "number",
    "incident_state",
    "active",
    "made_sla",
    "priority",
    "assignment_group",
    "assigned_to",
    "reassignment_count",
    "reopen_count",
    "sys_mod_count",
    "opened_at",
    "sys_updated_at",
    "resolved_at",
    "closed_at",
    "category"
]
# 2. Filter the dataframe for that specific sample_incident :------------>
incident_sample = (df[df["number"] == sample_incident][lifecycle_cols])
incident_sample






Toatal number of event-level history rows = 141712
unique_incidents in this datset = 24918  



,number,Total_Rows/Ticket
0,INC0019396,58
1,INC0044260,56
2,INC0005927,46
3,INC0020718,45
4,INC0011206,44
5,INC0007349,43
6,INC0025734,43
7,INC0012815,40
8,INC0003419,38
9,INC0020849,38


,index,rows_per_incident
0,count,24918.000000
1,mean,5.687134
2,std,3.677845
3,min,2.000000
4,25%,3.000000
5,50%,5.000000
6,75%,7.000000
7,max,58.000000


,number,incident_state,active,made_sla,priority,assignment_group,assigned_to,reassignment_count,reopen_count,sys_mod_count,opened_at,sys_updated_at,resolved_at,closed_at,category
81418,INC0019396,New,True,True,3 - Moderate,Group 66,?,0,0,0,14-04-2016 20:42,14-04-2016 20:42,12-05-2016 23:51,18-05-2016 00:07,Category 53
81419,INC0019396,New,True,True,3 - Moderate,Group 66,?,0,0,1,14-04-2016 20:42,14-04-2016 21:08,12-05-2016 23:51,18-05-2016 00:07,Category 53
81420,INC0019396,Resolved,True,True,3 - Moderate,Group 66,Resolver 175,0,0,6,14-04-2016 20:42,15-04-2016 09:20,12-05-2016 23:51,18-05-2016 00:07,Category 53
81421,INC0019396,Active,True,True,3 - Moderate,Group 66,Resolver 175,0,1,7,14-04-2016 20:42,15-04-2016 15:21,12-05-2016 23:51,18-05-2016 00:07,Category 53
81422,INC0019396,Active,True,True,3 - Moderate,Group 66,Resolver 175,0,1,9,14-04-2016 20:42,15-04-2016 15:21,12-05-2016 23:51,18-05-2016 00:07,Category 53
81423,INC0019396,Active,True,True,3 - Moderate,Group 66,Resolver 175,0,1,10,14-04-2016 20:42,15-04-2016 20:30,12-05-2016 23:51,18-05-2016 00:07,Category 53
81424,INC0019396,Active,True,True,3 - Moderate,Group 66,Resolver 175,0,1,11,14-04-2016 20:42,15-04-2016 20:30,12-05-2016 23:51,18-05-2016 00:07,Category 53
81425,INC0019396,Active,True,True,3 - Moderate,Group 66,Resolver 175,0,1,12,14-04-2016 20:42,16-04-2016 23:43,12-05-2016 23:51,18-05-2016 00:07,Category 53
81426,INC0019396,Active,True,True,3 - Moderate,Group 66,Resolver 175,0,1,16,14-04-2016 20:42,16-04-2016 23:43,12-05-2016 23:51,18-05-2016 00:07,Category 53
81427,INC0019396,Active,True,True,3 - Moderate,Group 54,?,1,1,18,14-04-2016 20:42,20-04-2016 11:12,12-05-2016 23:51,18-05-2016 00:07,Category 53


## Stage 2 Final Conclusion: Dataset Grain

The dataset is event-level/history-level, not incident-level.

Evidence:
- Total raw rows: 141,712
- Unique incident numbers: 24,918
- Average rows per incident: about 5.69
- Median rows per incident: 5
- Maximum rows for one incident: 58
- One repeated incident, INC0019396, has the most rows.
- Inside one incident, columns such as sys_updated_at, incident_state, assignment_group, assigned_to, reassignment_count, reopen_count, and sys_mod_count can change.

Important interpretation:
One raw row represents one history/update record of an incident ticket.
One incident ticket can appear multiple times.

Important warning:
Raw rows must not be counted directly as incidents.

Agent project meaning:
For future chatbot charts like SLA distribution, priority distribution, top assignment groups, reopened incident analysis, and reassignment analysis, I will later need one-row-per-incident data.
For lifecycle investigation questions, the raw event-level data may still be useful.

## Stage 3

In [32]:
agent_mvp_cols = [
    "number",
    "incident_state",
    "active",
    "made_sla",
    "priority",
    "impact",
    "urgency",
    "category",
    "subcategory",
    "assignment_group",
    "assigned_to",
    "reassignment_count",
    "reopen_count",
    "sys_mod_count",
    "opened_at",
    "sys_updated_at",
    "resolved_at",
    "closed_at",
    "problem_id",
    "change",
    "caused_by",
    "closed_code"
]
missing_mvp_cols = [col for col in agent_mvp_cols if col not in df.columns]

missing_mvp_cols

df_mvp = df[agent_mvp_cols]
df_mvp.head(10) 

#Why to choose these columns for the agent MVP?:-------------->
agent_mvp_column_notes = pd.DataFrame([
    {
        "column": "number",
        "business_meaning": "Unique incident ticket ID",
        
        "needs_cleaning_later": "Check duplicates through grain logic",
        "possible_agent_use_case": "Count unique incidents; filter one incident"
    },
    {
        "column": "incident_state",
        "business_meaning": "Current or historical state of the incident",
        
        "needs_cleaning_later": "Check inconsistent labels and final-state rule",
        "possible_agent_use_case": "Incident state distribution"
    },
    {
        "column": "active",
        "business_meaning": "Whether the incident is still active",
        
        "needs_cleaning_later": "Use latest row later",
        "possible_agent_use_case": "Open vs closed style questions"
    },
    {
        "column": "made_sla",
        "business_meaning": "Whether the incident met SLA",
        
        "needs_cleaning_later": "Use latest/final incident row later",
        "possible_agent_use_case": "SLA met vs breached distribution"
    },
    {
        "column": "priority",
        "business_meaning": "Business priority of the incident",
        
        "needs_cleaning_later": "Check values and final priority rule",
        "possible_agent_use_case": "Priority distribution"
    },
    {
        "column": "impact",
        "business_meaning": "Business impact level",
        
        "needs_cleaning_later": "Check labels and missing/fake missing values",
        "possible_agent_use_case": "Impact-based incident analysis"
    },
    {
        "column": "urgency",
        "business_meaning": "Urgency level of the incident",
        
        "needs_cleaning_later": "Check labels and missing/fake missing values",
        "possible_agent_use_case": "Urgency-based incident analysis"
    },
    {
        "column": "category",
        "business_meaning": "High-level incident category",
        
        "needs_cleaning_later": "Check fake missing and inconsistent labels",
        "possible_agent_use_case": "Top categories"
    },
    {
        "column": "subcategory",
        "business_meaning": "More specific incident category",
        
        "needs_cleaning_later": "Check fake missing and high cardinality",
        "possible_agent_use_case": "Top subcategories"
    },
    {
        "column": "assignment_group",
        "business_meaning": "Team responsible for the incident",
        
        "needs_cleaning_later": "Check fake missing and latest assignment rule",
        "possible_agent_use_case": "Top assignment groups"
    },
    {
        "column": "assigned_to",
        "business_meaning": "Person assigned to the incident",
        
        "needs_cleaning_later": "Check missing/fake missing; may be sensitive or too detailed",
        "possible_agent_use_case": "Investigation only, not first dashboard focus"
    },
    {
        "column": "reassignment_count",
        "business_meaning": "How many times the incident was reassigned",
        
        "needs_cleaning_later": "Use max value per incident later",
        "possible_agent_use_case": "Reassignment analysis"
    },
    {
        "column": "reopen_count",
        "business_meaning": "How many times the incident was reopened",
        
        "needs_cleaning_later": "Use max value per incident later",
        "possible_agent_use_case": "Reopened incident analysis"
    },
    {
        "column": "sys_mod_count",
        "business_meaning": "System modification count",
        
        "needs_cleaning_later": "Use max value per incident later",
        "possible_agent_use_case": "Complexity/activity signal"
    },
    {
        "column": "opened_at",
        "business_meaning": "When the incident was opened",
        
        "needs_cleaning_later": "Parse to datetime later",
        "possible_agent_use_case": "Incident volume over time; resolution time"
    },
    {
        "column": "sys_updated_at",
        "business_meaning": "When the row/history record was last updated",
        
        "needs_cleaning_later": "Parse to datetime and use for latest-row rule",
        "possible_agent_use_case": "Select final incident row later"
    },
    {
        "column": "resolved_at",
        "business_meaning": "When the incident was resolved",
        
        "needs_cleaning_later": "Parse to datetime; check missing/fake missing",
        "possible_agent_use_case": "Resolution time analysis"
    },
    {
        "column": "closed_at",
        "business_meaning": "When the incident was closed",
        
        "needs_cleaning_later": "Parse to datetime; check logical order",
        "possible_agent_use_case": "Closure and lifecycle analysis"
    },
    {
        "column": "problem_id",
        "business_meaning": "Linked problem record, if any",
        
        "needs_cleaning_later": "Likely fake missing; verify before RCA claims",
        "possible_agent_use_case": "Problem linkage investigation"
    },
    {
        "column": "change",
        "business_meaning": "Linked change record, if any",
        
        "needs_cleaning_later": "Likely fake missing; verify before change-related claims",
        "possible_agent_use_case": "Change-related incident investigation"
    },
    {
        "column": "caused_by",
        "business_meaning": "Possible cause reference",
        
        "needs_cleaning_later": "Likely fake missing; risky for root-cause claims",
        "possible_agent_use_case": "Investigation guidance, not confirmed RCA"
    },
    {
        "column": "closed_code",
        "business_meaning": "Closure reason/code",
        
        "needs_cleaning_later": "Check fake missing and categories",
        "possible_agent_use_case": "Closure reason analysis"
    }
])
agent_mvp_column_notes







[]

,number,incident_state,active,made_sla,priority,impact,urgency,category,subcategory,assignment_group,...,reopen_count,sys_mod_count,opened_at,sys_updated_at,resolved_at,closed_at,problem_id,change,caused_by,closed_code
0,INC0000045,New,True,True,3 - Moderate,2 - Medium,2 - Medium,Category 55,Subcategory 170,Group 56,...,0,0,29-02-2016 01:16,29-02-2016 01:23,29-02-2016 11:29,05-03-2016 12:00,?,?,?,code 5
1,INC0000045,Resolved,True,True,3 - Moderate,2 - Medium,2 - Medium,Category 55,Subcategory 170,Group 56,...,0,2,29-02-2016 01:16,29-02-2016 08:53,29-02-2016 11:29,05-03-2016 12:00,?,?,?,code 5
2,INC0000045,Resolved,True,True,3 - Moderate,2 - Medium,2 - Medium,Category 55,Subcategory 170,Group 56,...,0,3,29-02-2016 01:16,29-02-2016 11:29,29-02-2016 11:29,05-03-2016 12:00,?,?,?,code 5
3,INC0000045,Closed,False,True,3 - Moderate,2 - Medium,2 - Medium,Category 55,Subcategory 170,Group 56,...,0,4,29-02-2016 01:16,05-03-2016 12:00,29-02-2016 11:29,05-03-2016 12:00,?,?,?,code 5
4,INC0000047,New,True,True,3 - Moderate,2 - Medium,2 - Medium,Category 40,Subcategory 215,Group 70,...,0,0,29-02-2016 04:40,29-02-2016 04:57,01-03-2016 09:52,06-03-2016 10:00,?,?,?,code 5
5,INC0000047,Active,True,True,3 - Moderate,2 - Medium,2 - Medium,Category 40,Subcategory 215,Group 24,...,0,1,29-02-2016 04:40,29-02-2016 05:30,01-03-2016 09:52,06-03-2016 10:00,?,?,?,code 5
6,INC0000047,Active,True,True,3 - Moderate,2 - Medium,2 - Medium,Category 40,Subcategory 215,Group 24,...,0,2,29-02-2016 04:40,29-02-2016 05:33,01-03-2016 09:52,06-03-2016 10:00,?,?,?,code 5
7,INC0000047,Active,True,True,3 - Moderate,2 - Medium,2 - Medium,Category 40,Subcategory 215,Group 24,...,0,3,29-02-2016 04:40,29-02-2016 11:31,01-03-2016 09:52,06-03-2016 10:00,?,?,?,code 5
8,INC0000047,Active,True,True,3 - Moderate,2 - Medium,2 - Medium,Category 40,Subcategory 215,Group 24,...,0,4,29-02-2016 04:40,29-02-2016 11:32,01-03-2016 09:52,06-03-2016 10:00,?,?,?,code 5
9,INC0000047,Active,True,True,3 - Moderate,2 - Medium,2 - Medium,Category 40,Subcategory 215,Group 24,...,0,5,29-02-2016 04:40,01-03-2016 09:14,01-03-2016 09:52,06-03-2016 10:00,?,?,?,code 5


,column,business_meaning,needs_cleaning_later,possible_agent_use_case
0,number,Unique incident ticket ID,Check duplicates through grain logic,Count unique incidents; filter one incident
1,incident_state,Current or historical state of the incident,Check inconsistent labels and final-state rule,Incident state distribution
2,active,Whether the incident is still active,Use latest row later,Open vs closed style questions
3,made_sla,Whether the incident met SLA,Use latest/final incident row later,SLA met vs breached distribution
4,priority,Business priority of the incident,Check values and final priority rule,Priority distribution
5,impact,Business impact level,Check labels and missing/fake missing values,Impact-based incident analysis
6,urgency,Urgency level of the incident,Check labels and missing/fake missing values,Urgency-based incident analysis
7,category,High-level incident category,Check fake missing and inconsistent labels,Top categories
8,subcategory,More specific incident category,Check fake missing and high cardinality,Top subcategories
9,assignment_group,Team responsible for the incident,Check fake missing and latest assignment rule,Top assignment groups


## Stage 3 Conclusion: Agent MVP Column Understanding

In this stage, I selected the columns that are most relevant for the first version of the incident analytics agent.

`df[agent_mvp_cols]` was used to create a focused view of the raw dataset. The purpose was not to clean, transform, or export the data. The purpose was only to inspect the columns that are most useful for future business questions such as SLA distribution, priority distribution, incident state analysis, assignment group analysis, reopened incidents, and reassignment behavior.

`agent_mvp_column_notes` was created as a simple data dictionary. Its purpose is to explain why each selected column matters, what business meaning it has, and what data quality risks may need attention later.

Stage 3 finding:
The selected MVP columns give enough business context to support the future agent, but some columns still need validation in later stages before they can be used confidently.


## Stage 4: Fake Missing Value Check

In [33]:
df["cmdb_ci"].value_counts(dropna=False).head(20)
fake_tokens = ["?", "NA", "N/A", "None", "null", "Unknown", "-", ""]
fake_summary = []

for col in df.columns:
    fake_count = df[col].astype(str).str.strip().isin(fake_tokens).sum()
    
    fake_summary.append({
        "column": col,
        "fake_count": fake_count,
        "fake_percent": round((fake_count / len(df)) * 100, 2)
    })

fake_summary_df = pd.DataFrame(fake_summary)

fake_summary_df.sort_values(
    by="fake_percent",
    ascending=False
)

print(f"Name of the columns which have fake/missing values : {fake_summary_df[fake_summary_df['fake_count'] > 0]['column'].tolist()}")

cmdb_ci
?             141267
cmdb_ci 31        32
cmdb_ci 49        24
cmdb_ci 7         24
cmdb_ci 11        21
cmdb_ci 14        17
cmdb_ci 46        16
cmdb_ci 23        14
cmdb_ci 5         14
cmdb_ci 17        13
cmdb_ci 40        12
cmdb_ci 28        11
cmdb_ci 12        11
cmdb_ci 47        11
cmdb_ci 20        11
cmdb_ci 6         10
cmdb_ci 24        10
cmdb_ci 50        10
cmdb_ci 3          9
cmdb_ci 48         9
Name: count, dtype: int64

,column,fake_count,fake_percent
31,caused_by,141689,99.98
30,vendor,141468,99.83
19,cmdb_ci,141267,99.69
29,change,140721,99.30
28,problem_id,139417,98.38
10,sys_created_by,53076,37.45
11,sys_created_at,53076,37.45
18,symptom,32964,23.26
24,assigned_to,27496,19.40
23,assignment_group,14213,10.03


Name of the columns which have fake/missing values : ['caller_id', 'opened_by', 'sys_created_by', 'sys_created_at', 'location', 'category', 'subcategory', 'symptom', 'cmdb_ci', 'assignment_group', 'assigned_to', 'problem_id', 'change', 'vendor', 'caused_by', 'closed_code', 'resolved_by', 'resolved_at']


## Stage 4  : Final Conclusion :

Although df.info() may show many columns as non-null, this does not prove the data is complete.

A column like cmdb_ci is suspicious because in real ITSM data, not every incident is expected to have a known configuration item. If cmdb_ci appears non-null for every row, it may be because fake missing values such as ? are stored as text.

In this stage, Current understanding:
- caused_by, change, and problem_id are not ready for reliable analysis in raw form.
- These columns may still be useful after cleaning, but their quality must be checked carefully.
- RCA-related analysis should not be planned until these fields are cleaned and validated.

Cleaning questions for the next notebook:
1. Should ? be converted into real missing values?
2. Should caused_by, change, and problem_id be kept, dropped, or marked as low-confidence fields?
3. Should these columns be used only for investigation context instead of confirmed root-cause analysis?
4. Should missingness itself become a data quality finding?

## Stage 5: Key Categorical Distribution Checks

In [34]:
df["sys_updated_at"].value_counts(dropna=False).head(20)

df["caller_id"].value_counts(dropna=False).head(20)

df["priority"].value_counts(dropna=False)

df["made_sla"].value_counts(dropna=False)


df["incident_state"].value_counts(dropna=False)
df[df["incident_state"] == "-100"][
    ["number", "incident_state", "sys_updated_at", "opened_at", "closed_at" , "priority", "assignment_group"]
]


df["assignment_group"].value_counts(dropna=False).head(10)
print(f"Total number of unique assignment groups including place holder values E.g.--->'?' : {df['assignment_group'].nunique(dropna=False)} \n")

df["category"].value_counts(dropna=False).head(20)
print(f"Total number of unique category : {df['category'].nunique(dropna=False)} & number of rows occupied by fake placeholders in category column : {df['category'].astype(str).str.strip().isin(fake_tokens).sum()} \n") ## AS in stage 4 we defined fake_tokens = ["?", "NA", "N/A", "None", "null", "Unknown", "-", ""]


df["subcategory"].value_counts(dropna=False).head(20)
print(f"Total number of unique subcategory: {df['subcategory'].nunique(dropna=False)} & Fake placeholder rows in subcategory: {df['subcategory'].astype(str).str.strip().isin(fake_tokens).sum()}")


df["impact"].value_counts(dropna=False)

df["urgency"].value_counts(dropna=False)





sys_updated_at
24-03-2016 18:40    420
24-03-2016 19:00    405
24-03-2016 18:51    400
24-03-2016 19:01    393
24-03-2016 18:41    386
24-03-2016 18:50    369
24-03-2016 18:59    348
24-03-2016 19:02    270
23-05-2016 11:07    214
24-03-2016 18:52    140
17-05-2016 16:07     76
03-05-2016 11:07     72
28-03-2016 15:59     67
03-04-2016 16:59     67
27-03-2016 11:59     66
21-05-2016 10:07     65
30-04-2016 10:07     63
05-04-2016 10:59     63
13-04-2016 15:59     63
02-05-2016 12:07     63
Name: count, dtype: int64

caller_id
Caller 1904    1425
Caller 290      791
Caller 4514     716
Caller 1441     322
Caller 298      293
Caller 3763     270
Caller 93       239
Caller 1531     231
Caller 4414     224
Caller 3160     220
Caller 2471     219
Caller 90       219
Caller 3479     207
Caller 1270     207
Caller 363      204
Caller 3870     201
Caller 1517     198
Caller 707      196
Caller 5093     194
Caller 994      191
Name: count, dtype: int64

priority
3 - Moderate    132452
4 - Low           4030
2 - High          2972
1 - Critical      2258
Name: count, dtype: int64

made_sla
True     132497
False      9215
Name: count, dtype: int64

incident_state
Active                38716
New                   36407
Resolved              25751
Closed                24985
Awaiting User Info    14642
Awaiting Vendor         707
Awaiting Problem        461
Awaiting Evidence        38
-100                      5
Name: count, dtype: int64

,number,incident_state,sys_updated_at,opened_at,closed_at,priority,assignment_group
115641,INC0028089,-100,16-06-2016 12:02,10-05-2016 22:09,21-06-2016 13:07,3 - Moderate,Group 64
115642,INC0028089,-100,16-06-2016 12:03,10-05-2016 22:09,21-06-2016 13:07,3 - Moderate,Group 64
124507,INC0030413,-100,15-06-2016 17:15,17-05-2016 14:42,26-06-2016 18:00,3 - Moderate,Group 17
124508,INC0030413,-100,15-06-2016 17:21,17-05-2016 14:42,26-06-2016 18:00,3 - Moderate,Group 17
124509,INC0030413,-100,21-06-2016 17:27,17-05-2016 14:42,26-06-2016 18:00,3 - Moderate,Group 17


assignment_group
Group 70    43474
?           14213
Group 25     7679
Group 24     6752
Group 20     6170
Group 39     4728
Group 23     4228
Group 27     3817
Group 28     3638
Group 73     3304
Name: count, dtype: int64

Total number of unique assignment groups including place holder values E.g.--->'?' : 79 



category
Category 26    18453
Category 42    15977
Category 53    15968
Category 46    13324
Category 23     7779
Category 9      7365
Category 32     7273
Category 37     6584
Category 57     6532
Category 20     5506
Category 61     5168
Category 24     4561
Category 34     3946
Category 40     3760
Category 45     3619
Category 51     2390
Category 28     2256
Category 35     2037
Category 44     1798
Category 19     1592
Name: count, dtype: int64

Total number of unique category : 59 & number of rows occupied by fake placeholders in category column : 78 



subcategory
Subcategory 174    35785
Subcategory 223    15598
Subcategory 175    10496
Subcategory 164     7880
Subcategory 9       4124
Subcategory 275     3643
Subcategory 135     3467
Subcategory 170     3349
Subcategory 36      2917
Subcategory 75      2113
Subcategory 125     1947
Subcategory 3       1788
Subcategory 43      1674
Subcategory 150     1659
Subcategory 28      1565
Subcategory 303     1558
Subcategory 114     1546
Subcategory 103     1517
Subcategory 123     1425
Subcategory 154     1170
Name: count, dtype: int64

Total number of unique subcategory: 255 & Fake placeholder rows in subcategory: 111


impact
2 - Medium    134335
3 - Low         3886
1 - High        3491
Name: count, dtype: int64

urgency
2 - Medium    134094
1 - High        4020
3 - Low         3598
Name: count, dtype: int64

## Stage 5 Conclusion: Categorical Distribution Understanding

In this stage, I checked the main categorical columns needed for future analysis: made_sla, priority, incident_state, assignment_group, category, subcategory, impact, and urgency.

The key learning is that value_counts() helps me understand how often each category appears in the raw dataset.

Important findings:
- sys_updated_at, caller_id , priority, made_sla, impact, and urgency are useful categorical fields with readable values.
- incident_state is mostly readable, but the value -100 appears and should be investigated later.
- assignment_group is useful for team/ownership analysis, but it has 79 unique values and also contains ? placeholders.
- category has 59 unique values and only a small number of fake placeholder rows.
- subcategory has 255 unique values, so future analysis may need Top 10 + Others logic.
- category and subcategory labels are anonymized, so they are useful for pattern analysis but not full business explanation unless a mapping table exists.

Important caution:
These distributions are based on raw history rows, not unique incident tickets. Since the dataset is event-level, final incident-level distributions may change after aggregation.

This stage only identifies category patterns and data understanding issues. It does not clean or transform the data.

## Stage 6: Datetime Parsing Sanity Check.

In [35]:
DT_cols = [
    "opened_at",
    "sys_created_at",
    "sys_updated_at",
    "resolved_at",
    "closed_at"
]

# Use .apply() to run to_datetime on each column individually
DT_at_test = df[DT_cols].apply(pd.to_datetime, errors="coerce", dayfirst=True)

DT_at_test.head(10)
print(f"Number of missing or unparseable datetime values per column: \n {DT_at_test.isna().sum()}")


unparseable_sys_created_at = DT_at_test["sys_created_at"].isna()
df.loc[unparseable_sys_created_at,"sys_created_at"].value_counts(dropna=False)

unparseable_resolved_at = DT_at_test["resolved_at"].isna()
df.loc[unparseable_resolved_at,"resolved_at"].value_counts(dropna=False)


opened_after_resolved = (
    DT_at_test["opened_at"].notna()
    & DT_at_test["resolved_at"].notna()
    & (DT_at_test["opened_at"] > DT_at_test["resolved_at"])
)
opened_after_resolved.sum()


resolved_after_closed = (
    DT_at_test["resolved_at"].notna()
    & DT_at_test["closed_at"].notna()
    & (DT_at_test["resolved_at"] > DT_at_test["closed_at"])
)
resolved_after_closed.sum()

sys_created_before_opened = (
    DT_at_test["sys_created_at"].notna()
    & DT_at_test["opened_at"].notna()
    & (DT_at_test["sys_created_at"] < DT_at_test["opened_at"])
)
sys_created_before_opened.sum()


sys_updated_before_opened = (
        DT_at_test["sys_updated_at"].notna()
        & DT_at_test["opened_at"].notna()
        & (DT_at_test["sys_updated_at"] < DT_at_test["opened_at"])
)
sys_updated_before_opened.sum()







,opened_at,sys_created_at,sys_updated_at,resolved_at,closed_at
0,2016-02-29 01:16:00,2016-02-29 01:23:00,2016-02-29 01:23:00,2016-02-29 11:29:00,2016-03-05 12:00:00
1,2016-02-29 01:16:00,2016-02-29 01:23:00,2016-02-29 08:53:00,2016-02-29 11:29:00,2016-03-05 12:00:00
2,2016-02-29 01:16:00,2016-02-29 01:23:00,2016-02-29 11:29:00,2016-02-29 11:29:00,2016-03-05 12:00:00
3,2016-02-29 01:16:00,2016-02-29 01:23:00,2016-03-05 12:00:00,2016-02-29 11:29:00,2016-03-05 12:00:00
4,2016-02-29 04:40:00,2016-02-29 04:57:00,2016-02-29 04:57:00,2016-03-01 09:52:00,2016-03-06 10:00:00
5,2016-02-29 04:40:00,2016-02-29 04:57:00,2016-02-29 05:30:00,2016-03-01 09:52:00,2016-03-06 10:00:00
6,2016-02-29 04:40:00,2016-02-29 04:57:00,2016-02-29 05:33:00,2016-03-01 09:52:00,2016-03-06 10:00:00
7,2016-02-29 04:40:00,2016-02-29 04:57:00,2016-02-29 11:31:00,2016-03-01 09:52:00,2016-03-06 10:00:00
8,2016-02-29 04:40:00,2016-02-29 04:57:00,2016-02-29 11:32:00,2016-03-01 09:52:00,2016-03-06 10:00:00
9,2016-02-29 04:40:00,2016-02-29 04:57:00,2016-03-01 09:14:00,2016-03-01 09:52:00,2016-03-06 10:00:00


Number of missing or unparseable datetime values per column: 
 opened_at             0
sys_created_at    53076
sys_updated_at        0
resolved_at        3141
closed_at             0
dtype: int64


sys_created_at
?    53076
Name: count, dtype: int64

resolved_at
?    3141
Name: count, dtype: int64

np.int64(0)

np.int64(0)

np.int64(34)

np.int64(5)

## Stage 6 Conclusion: Datetime Parsing Sanity Check

In this stage, I checked whether the main date columns can be converted from text/object values into real pandas datetime values.

The datetime columns checked were:
- opened_at
- sys_created_at
- sys_updated_at
- resolved_at
- closed_at

Key findings:
- opened_at, sys_updated_at, and closed_at parsed successfully with 0 failed values.
- sys_created_at had 53,076 unparseable values because the raw value was ?.
- resolved_at had 3,141 unparseable values because the raw value was ?.
- The date format appears to be day-month-year hour:minute, so future parsing should use dayfirst=True or an explicit datetime format.
- No rows were found where opened_at is after resolved_at.
- No rows were found where resolved_at is after closed_at.
- 34 rows were found where sys_created_at is before opened_at, representing minor chronological inconsistencies.
- 5 rows were found where sys_updated_at is before opened_at, representing minor chronological inconsistencies.

Data understanding:
The main datetime issue is not broken date formatting. The main issue is fake missing datetime placeholders, especially in sys_created_at and resolved_at.



## Stage 7: One Incident Lifecycle Trace

In [36]:
sample_incident = "INC0019396"

incident_lifecycle = df[df["number"] == sample_incident].copy()

incident_lifecycle.shape

incident_lifecycle["sys_updated_at_parsed"] = pd.to_datetime(
    incident_lifecycle["sys_updated_at"] , errors="coerce" , dayfirst=True
)
incident_lifecycle["sys_updated_at_parsed"].isna().sum()


#Now display only the useful lifecycle columns:---->
incident_lifecycle_cols = [
        "number",
        "incident_state",
        "active",
        "made_sla",
        "priority",
        "assignment_group",
        "assigned_to",
        "reassignment_count",
        "reopen_count",
        "sys_mod_count",
        "opened_at",
        "sys_updated_at",
        "resolved_at",
        "closed_at"
    ]

incident_lifecycle_cols_df = incident_lifecycle.sort_values("sys_updated_at_parsed")

incident_lifecycle_cols_df[incident_lifecycle_cols].nunique(dropna=False).sort_values(ascending=False) 

incident_lifecycle_cols_df[incident_lifecycle_cols]



(58, 37)

np.int64(0)

sys_mod_count         58
sys_updated_at        40
reassignment_count    11
reopen_count           9
incident_state         5
assigned_to            5
assignment_group       4
made_sla               2
active                 2
number                 1
priority               1
opened_at              1
resolved_at            1
closed_at              1
dtype: int64

,number,incident_state,active,made_sla,priority,assignment_group,assigned_to,reassignment_count,reopen_count,sys_mod_count,opened_at,sys_updated_at,resolved_at,closed_at
81418,INC0019396,New,True,True,3 - Moderate,Group 66,?,0,0,0,14-04-2016 20:42,14-04-2016 20:42,12-05-2016 23:51,18-05-2016 00:07
81419,INC0019396,New,True,True,3 - Moderate,Group 66,?,0,0,1,14-04-2016 20:42,14-04-2016 21:08,12-05-2016 23:51,18-05-2016 00:07
81420,INC0019396,Resolved,True,True,3 - Moderate,Group 66,Resolver 175,0,0,6,14-04-2016 20:42,15-04-2016 09:20,12-05-2016 23:51,18-05-2016 00:07
81421,INC0019396,Active,True,True,3 - Moderate,Group 66,Resolver 175,0,1,7,14-04-2016 20:42,15-04-2016 15:21,12-05-2016 23:51,18-05-2016 00:07
81422,INC0019396,Active,True,True,3 - Moderate,Group 66,Resolver 175,0,1,9,14-04-2016 20:42,15-04-2016 15:21,12-05-2016 23:51,18-05-2016 00:07
81423,INC0019396,Active,True,True,3 - Moderate,Group 66,Resolver 175,0,1,10,14-04-2016 20:42,15-04-2016 20:30,12-05-2016 23:51,18-05-2016 00:07
81424,INC0019396,Active,True,True,3 - Moderate,Group 66,Resolver 175,0,1,11,14-04-2016 20:42,15-04-2016 20:30,12-05-2016 23:51,18-05-2016 00:07
81425,INC0019396,Active,True,True,3 - Moderate,Group 66,Resolver 175,0,1,12,14-04-2016 20:42,16-04-2016 23:43,12-05-2016 23:51,18-05-2016 00:07
81426,INC0019396,Active,True,True,3 - Moderate,Group 66,Resolver 175,0,1,16,14-04-2016 20:42,16-04-2016 23:43,12-05-2016 23:51,18-05-2016 00:07
81427,INC0019396,Active,True,True,3 - Moderate,Group 54,?,1,1,18,14-04-2016 20:42,20-04-2016 11:12,12-05-2016 23:51,18-05-2016 00:07


## Stage 7 Conclusion: One Incident Lifecycle Trace

I inspected one repeated incident, INC0019396, which has 58 raw history rows.

The lifecycle trace shows that one incident can appear many times because the dataset stores updates over time.

Key findings:
- opened_at stayed stable for this incident, meaning the ticket was opened once.
- sys_updated_at changed across the lifecycle, showing different update moments.
- sys_mod_count changed in every row, making it a strong lifecycle/activity field.
- reassignment_count, reopen_count, assignment_group, assigned_to, active, made_sla, and incident_state changed during the lifecycle.
- number, priority, opened_at, resolved_at, and closed_at stayed stable for this sampled incident.

Data understanding:
This confirms that the raw dataset is event-level/history-level. Some fields describe the incident itself, while other fields describe changes during the incident lifecycle.

This stage helps plan future aggregation rules, but does not create the final incident-level dataset.

## Stage 8.1 : Incident Lifecycle Understanding — How Tickets Move Over Time

In [37]:
df_lifecycle = df.copy()

column_change_summary = []

for col in df_lifecycle.columns:
    incidents_where_changed = (
        df_lifecycle.groupby("number")[col]
                    .nunique(dropna=False)
                    .gt(1)
                    .sum()
    )

    column_change_summary.append({
        "column": col,
        "incidents_where_column_changed": incidents_where_changed,
        "%_of_incidents_changed": round(
            incidents_where_changed / df_lifecycle["number"].nunique() * 100,
            2
        )
    })

column_change_summary_df = (
    pd.DataFrame(column_change_summary)
    .sort_values("incidents_where_column_changed", ascending=False)
)

column_change_summary_df



,column,incidents_where_column_changed,%_of_incidents_changed
1,incident_state,24918,100.00
2,active,24918,100.00
5,sys_mod_count,24918,100.00
13,sys_updated_at,24909,99.96
12,sys_updated_by,23379,93.82
26,u_priority_confirmation,19124,76.75
3,reassignment_count,11369,45.63
23,assignment_group,11344,45.53
6,made_sla,9114,36.58
24,assigned_to,8949,35.91


## Stage 8.2 : Validate cumulative counters

In [38]:
df_lifecycle["sys_updated_at_parsed"] = pd.to_datetime(df_lifecycle["sys_updated_at"], errors="coerce", dayfirst=True)

def has_decreasing_counter(group, counter_col):
    group = group.sort_values(["sys_updated_at_parsed", "sys_mod_count"])
    return (group[counter_col].diff().dropna() < 0).any()


counter_monotonicity_results = []

for counter_col in ["reassignment_count", "reopen_count", "sys_mod_count"]:
    incidents_with_decrease = (
        df_lifecycle
        .groupby("number")
        .apply(lambda group: has_decreasing_counter(group, counter_col))
        .sum()
    )

    counter_monotonicity_results.append({
        "counter_column": counter_col,
        "incidents_with_decreasing_counter": incidents_with_decrease
    })

pd.DataFrame(counter_monotonicity_results)



,counter_column,incidents_with_decreasing_counter
0,reassignment_count,0
1,reopen_count,0
2,sys_mod_count,5


## Stage 8.3: Validating the Final State (Latest-Row) Extraction Logic

In [39]:
df_latest_test = df.copy()

df_latest_test["sys_updated_at_parsed"] = pd.to_datetime(
    df_latest_test["sys_updated_at"],
    errors="coerce",
    dayfirst=True
)

latest_rows = (
    df_latest_test
    .sort_values(["number", "sys_updated_at_parsed", "sys_mod_count"])
    .groupby("number")
    .tail(1)
)

print("Latest-row dataset shape:", latest_rows.shape)
print("Unique incidents:", latest_rows["number"].nunique())

latest_rows["incident_state"].value_counts(dropna=False)
latest_rows["active"].value_counts(dropna=False)
latest_rows["made_sla"].value_counts(dropna=False)

Latest-row dataset shape: (24918, 38)
Unique incidents: 24918


incident_state
Closed    24918
Name: count, dtype: int64

active
False    24918
Name: count, dtype: int64

made_sla
True     15803
False     9115
Name: count, dtype: int64

# Stage 8 Conclusion: 


### Aggregation Rule Planning

The raw dataset is event-level/history-level, meaning one incident ticket can appear across multiple rows as it changes over time.

For future incident-level analysis, the data will later need to be converted into one row per incident. This stage does not create that final dataset. It only defines the aggregation rules that should be used later in Data cleaning notebook or the agent-ready preparation step.

Aggregation is needed because raw history rows should not be counted directly as incidents. For example, one incident may have many updates, reassignment changes, state changes, and system modifications. Without aggregation, incident volume, SLA distribution, assignment group workload, reopen counts, and reassignment analysis could be misleading.

### Planned Aggregation Rules

| Column               | Aggregation Rule                                 | Reason                                                                                                                         |
| -------------------- | ------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------ |
| `number`             | `group key`                                      | Identifies the unique incident ticket. All event/history rows with the same number belong to the same incident.                |
| `opened_at`          | `earliest valid value`                           | Represents when the ticket was first opened. This is the start point of the incident lifecycle.                                |
| `sys_updated_at`     | `latest valid value`                             | Represents the latest update timestamp for the incident history. Useful for identifying the final/latest row.                  |
| `resolved_at`        | `latest valid value`                             | Represents when the incident was marked resolved. The latest valid value should represent the final resolved timestamp.        |
| `closed_at`          | `latest valid value`                             | Represents when the incident was officially closed. The latest valid value should represent the final closure timestamp.       |
| `incident_state`     | `value from latest sys_updated_at row`           | Represents the latest known state of the incident, such as Active, Resolved, or Closed.                                        |
| `active`             | `value from latest sys_updated_at row`           | Represents whether the incident was active in the latest known ticket state.                                                   |
| `made_sla`           | `value from latest sys_updated_at row`           | Represents the latest known SLA status for the incident.                                                                       |
| `priority`           | `value from latest sys_updated_at row`           | Represents the latest known priority after all ticket updates.                                                                 |
| `impact`             | `value from latest sys_updated_at row`           | Represents the latest known business impact level.                                                                             |
| `urgency`            | `value from latest sys_updated_at row`           | Represents the latest known urgency level.                                                                                     |
| `assignment_group`   | `value from latest sys_updated_at row`           | Represents the latest known owning/support group for the incident.                                                             |
| `assigned_to`        | `value from latest sys_updated_at row`           | Represents the latest known assigned person for the incident.                                                                  |
| `reassignment_count` | `maximum value`                                  | This is an increasing counter. The maximum value represents the final total number of reassignments.                           |
| `reopen_count`       | `maximum value`                                  | This is an increasing counter. The maximum value represents the final total number of reopens.                                 |
| `sys_mod_count`      | `maximum value`                                  | This is an increasing counter. The maximum value represents the final total number of system updates/modifications.            |
| `category`           | `latest or most common value`                    | Category is a classification field. It may stay stable or change, so the best rule should be validated later.                  |
| `subcategory`        | `latest or most common value`                    | Subcategory is a more detailed classification field. It may stay stable or change, so the best rule should be validated later. |
| `problem_id`         | `latest valid value after fake-missing handling` | This field has many `?` placeholders, so fake missing values must be handled before selecting a useful value.                  |
| `change`             | `latest valid value after fake-missing handling` | This field has many `?` placeholders, so fake missing values must be handled before selecting a useful value.                  |
| `caused_by`          | `latest valid value after fake-missing handling` | This field has very high fake-missing risk, so it needs careful cleaning and validation before RCA-style analysis.             |
| `closed_code`        | `latest valid value after fake-missing handling` | Closure code may contain fake missing values, so the latest valid value should be selected after cleaning.                     |



### Data Understanding Decision

Before creating `incidents_agent_ready.csv`, the next notebook should handle:

* fake missing values such as `?`
* datetime conversion
* validation of category and subcategory behavior
* latest-row selection using `sys_updated_at`
* proper incident-level aggregation by `number`
* handle suspicious values
* prepare validated columns for incident-level aggregation
* create a cleaned foundation for later EDA and agent-ready output

### Assumption Validation Table

This stage validates the main assumptions discovered during data understanding. The goal is to separate what is strongly supported by the data from what still needs validation later.

| Assumption | Evidence from data | Status | Risk if wrong |
| :--- | :--- | :--- | :--- |
| Dataset is event-level/history-level. | The dataset has 141,712 raw rows but only 24,918 unique incident numbers. One sampled incident, `INC0019396`, has 58 rows. | Validated | Raw rows could be incorrectly counted as incidents, inflating KPIs. |
| `number` is the incident ID. | Filtering by `number` returns the lifecycle rows for one ticket. `INC0019396` showed multiple history rows for the same incident. | Validated | Grouping by the wrong field would create incorrect incident-level records. |
| `?` represents fake missing values. | `cmdb_ci`, `caused_by`, `change`, `problem_id`, `sys_created_at`, and `resolved_at` contain `?` values where real values are expected. | Validated | Missingness would be hidden, and data quality would look better than it really is. |
| Latest `sys_updated_at` row can represent the final/latest incident state. | Stage 8.1 proved that `sys_updated_at` changes in 99.96% of incidents, making it a highly reliable chronological tracker for ordering states. | Validated | If latest-row logic is wrong, final state, SLA status, owner, or priority may be selected incorrectly. |
| Counters like `reassignment_count`, `reopen_count`, and `sys_mod_count` should use max. | Stage 8.2 proved `reassignment_count` and `reopen_count` never decrease chronologically. `sys_mod_count` only decreased in 5 rare anomalies out of 24,918 incidents. | Validated | Wrong aggregation could exaggerate reassignments, reopens, or system updates. |
| Some columns change inside the same incident. | Stage 8.1 proved that `incident_state`, `active`, `sys_mod_count` change 100% of the time, and `assignment_group` changes 45% of the time. | Validated | Treating changing fields as stable could create wrong final incident records. |
| Some columns are stable incident attributes. | Stage 8.1 proved that `number`, `opened_at`, `sys_created_at`, `resolved_at`, and `closed_at` change 0.00% of the time across all incidents. | Validated | If a supposedly stable field changes in other incidents, the aggregation rule may need adjustment. |
| Dataset can support insight and investigation guidance, but not confirmed root cause. | RCA-related fields are mostly fake missing: `caused_by` is 99.98% fake missing, `change` is 99.30%, and `problem_id` is 98.38%. | Validated | The project may overclaim root-cause findings that the raw data cannot support. |                      |


